### PR TITLE
[FLINK-16822] [sql-client] `table.xx` property set from CLI should also be set into TableEnvironment's TableConfig

### PR DESCRIPTION
**Pull request by godfreyhe**
## What is the purpose of the change

*The config set by SET command does not work, the reason is `table.xx` property set from CLI does not be set into TableEnvironment's TableConfig.*

## Brief change log

  - *Update table config of SessionState when calling setSessionProperty method*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added testSetSessionProperties in LocalExecutorITCase to verify the bug*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

### CODE REVIEW

**Comment by wuchong** [comment]
Can this also fix runtime environment configurations?

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java [hunk 1]
```diff
@@ -281,13 +283,18 @@ public void setSessionProperty(String sessionId, String key, String value) throw
 		ExecutionContext<?> context = getExecutionContext(sessionId);
 		Environment env = context.getEnvironment();
 		Environment newEnv = Environment.enrich(env, ImmutableMap.of(key, value), ImmutableMap.of());
+		ExecutionContext.SessionState sessionState = context.getSessionState();
+		// update table config
+		newEnv.getConfiguration().asMap().forEach((k, v) ->
+				sessionState.config.getConfiguration().setString(k, v));
```
Comment (wuchong) [inline flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java]: Will this influence the `RESET` command which should clear all session properties?

Comment (godfreyhe) [reply flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java]: good catch, we should store the original `SessionState` for resetting all properties to initial state.

**Comment by godfreyhe** [comment]
> Can this also fix runtime environment configurations? Currently, SET command can't set a runtime configuration because we ignore them in `Environment#enrich`, e.g. `SET pipeline.object-reuse=true` doesn't work.

nope, I think it's a new feature. Currently only few runtime configuration are supported, I think we can create a jira to support it.

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/ExecutionContext.java [hunk 2]
```diff
@@ -771,28 +772,24 @@ public Builder sessionState(SessionState sessionState) {
...
 			this.functionCatalog = functionCatalog;
 		}
 
 		public static SessionState of(
-				TableConfig config,
 				CatalogManager catalogManager,
 				ModuleManager moduleManager,
 				FunctionCatalog functionCatalog) {
-			return new SessionState(config, catalogManager, moduleManager, functionCatalog);
+			return new SessionState( catalogManager, moduleManager, functionCatalog);
```
Comment (wuchong) [inline flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/ExecutionContext.java:792]:
```suggestion
			return new SessionState(catalogManager, moduleManager, functionCatalog);
```
remove space.

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java [hunk 3]
```diff
@@ -281,13 +283,15 @@ public void setSessionProperty(String sessionId, String key, String value) throw
 		ExecutionContext<?> context = getExecutionContext(sessionId);
 		Environment env = context.getEnvironment();
 		Environment newEnv = Environment.enrich(env, ImmutableMap.of(key, value), ImmutableMap.of());
+		ExecutionContext.SessionState sessionState = context.getSessionState();
```
Comment (wuchong) [inline flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java:286]: Do we need this change? It seems they are equivalent.

Comment (godfreyhe) [reply flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java:286]: I revert this

**Comment by godfreyhe** [comment]
> `TableConfig`can

yes, it's hard to handle "RESET" command, we must remember which configs were set before and make sure TableConfig can also be set to init state.

**Comment by wuchong** [comment]
Could you open a new pull request for branch 1.9 too? @godfreyhe

**Comment by godfreyhe** [comment]
> branch

https://github.com/apache/flink/pull/11763, pr for 1.10. 
we are planing to publish a new release for 1.9 ?

**Comment by wuchong** [comment]
@godfreyhe , sorry, I wanted to say 1.10 branch.

**Pull request merged**